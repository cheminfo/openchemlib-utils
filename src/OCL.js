let OCL;

export function getOCL() {
  if (!OCL) {
    throw new Error('OCL has to be initialized using initOCL(OCL)');
  }
  return OCL;
}
/**
 * 
 * @param {OCL} newOCL 
 * @param {object} [options={}]
 * @param {boolean} [options.keepExisting=false] Will keep current OCL if we initialize more than once
 */
export function initOCL(newOCL, options = {}) {
  const { keepExisting = false } = options;
  if (OCL && keepExisting) return;
  if (OCL && OCL !== newOCL) {
    throw new Error('OCL-utils was already initialized with a different OCL');
  }
  OCL = newOCL;
}
