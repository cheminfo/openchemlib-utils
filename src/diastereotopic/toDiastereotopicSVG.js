import { getDiastereotopicAtomIDsAndH } from './getDiastereotopicAtomIDsAndH';

/**
 * Returns a SVG
 * @param {*} molecule
 * @param {*} [options={}]
 */
export function toDiastereotopicSVG(molecule, options = {}) {
  let {
    width = 300,
    height = 200,
    prefix = 'ocl',
    heavyAtomHydrogen = false,
  } = options;
  let svg = options.svg;
  let diaIDs = [];

  let hydrogenInfo = {};
  getDiastereotopicAtomIDsAndH(molecule).forEach((line) => {
    hydrogenInfo[line.oclID] = line;
  });

  if (heavyAtomHydrogen) {
    for (let i = 0; i < molecule.getAtoms(); i++) {
      diaIDs.push([]);
    }
    let groupedDiaIDs = molecule.getGroupedDiastereotopicAtomIDs();
    groupedDiaIDs.forEach((diaID) => {
      if (
        hydrogenInfo[diaID.oclID] &&
        hydrogenInfo[diaID.oclID].nbHydrogens > 0
      ) {
        diaID.atoms.forEach((atom) => {
          hydrogenInfo[diaID.oclID].hydrogenOCLIDs.forEach((id) => {
            if (!diaIDs[atom * 1].includes(id)) diaIDs[atom].push(id);
          });
        });
      }
    });
  } else {
    diaIDs = molecule.getDiastereotopicAtomIDs().map((a) => [a]);
  }

  if (!svg) svg = molecule.toSVG(width, height, prefix);

  svg = svg.replace(/Atom:[0-9]+"/g, (value) => {
    let atom = value.replace(/[^0-9]/g, '');
    return `${value} data-diaid="${diaIDs[atom].join(',')}"`;
  });

  return svg;
}
