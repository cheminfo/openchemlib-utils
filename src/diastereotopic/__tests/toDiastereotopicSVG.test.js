import OCL from 'openchemlib';

import { toDiastereotopicSVG } from '../toDiastereotopicSVG';

describe('toDiastereotopicSVG', () => {
  it('CCC', () => {
    let molecule = OCL.Molecule.fromSmiles('CCC');
    let svg = toDiastereotopicSVG(molecule);
    expect(svg).toHaveLength(996);
    expect(svg).toMatchSnapshot();
  });
});
