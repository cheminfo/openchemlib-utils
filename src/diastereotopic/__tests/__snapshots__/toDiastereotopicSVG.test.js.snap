// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toDiastereotopicSVG CCC 1`] = `
"<svg id=\\"ocl\\" xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\" width=\\"300px\\" height=\\"200px\\" viewBox=\\"0 0 300 200\\">
  <style> #ocl { pointer-events:none; } #ocl .event  { pointer-events:all; } line { stroke-linecap:round; } polygon { stroke-linejoin:round; } </style>
  <line x1=\\"170.78\\" y1=\\"106\\" x2=\\"150\\" y2=\\"94\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line x1=\\"150\\" y1=\\"94\\" x2=\\"129.22\\" y2=\\"106\\" stroke=\\"rgb(0,0,0)\\" stroke-width=\\"1.44\\" />
  <line id=\\"ocl:Bond:0\\" class=\\"event\\" x1=\\"170.78\\" y1=\\"106\\" x2=\\"150\\" y2=\\"94\\" stroke-width=\\"8\\" opacity=\\"0\\" />
  <line id=\\"ocl:Bond:1\\" class=\\"event\\" x1=\\"150\\" y1=\\"94\\" x2=\\"129.22\\" y2=\\"106\\" stroke-width=\\"8\\" opacity=\\"0\\" />
  <circle id=\\"ocl:Atom:0\\" data-diaid=\\"eM@Df\`Xb\`RP\\\\Jh\\" class=\\"event\\" cx=\\"170.78\\" cy=\\"106\\" r=\\"8\\" opacity=\\"0\\" />
  <circle id=\\"ocl:Atom:1\\" data-diaid=\\"eM@HzAbJC}IApj\`\\" class=\\"event\\" cx=\\"150\\" cy=\\"94\\" r=\\"8\\" opacity=\\"0\\" />
  <circle id=\\"ocl:Atom:2\\" data-diaid=\\"eM@Df\`Xb\`RP\\\\Jh\\" class=\\"event\\" cx=\\"129.22\\" cy=\\"106\\" r=\\"8\\" opacity=\\"0\\" />
</svg>"
`;
